File: vbox.cpp
/*
 * gfdgd xi
 */
#include "vbox.h"
#include "command.h"
#include <QMessageBox>

vbox::vbox(QString name, QString managerPath) {
    this->name = name;
    this->managerPath = managerPath;
    //Command command = Command();
    this->vboxVersion = Command().GetCommand("'" + managerPath + "' -v");
}

int vbox::Create(QString type){
    system(("\"" + managerPath + "\" createvm --name \""
                   + name + "\" --ostype \"" + type +
                   "\" --register").toLatin1());
    return system(("\"" + managerPath + "\" modifyvm \""
                   + name + "\" --ostype \"" + type +
                   "\" ").toLatin1());
    //vboxmanage modifyvm testvm --ostype
}
int vbox::CreateDisk(QString path, int size){
    return system(("\"" + managerPath + "\" createvdi --filename \"" + path + "\" --size \"" + QString::number(size) + "\"").toLatin1());
}
int vbox::CreateDiskControl(QString controlName){
    return system(("\"" + managerPath + "\" storagectl \"" + name + "\" --name \"" + controlName + "\" --add ide").toLatin1());
}
int vbox::MountDisk(QString diskPath, QString controlName, int port, int device){
    return system(("\"" + managerPath + "\" storageattach \"" + name +
                   "\" --storagectl \"" + controlName + "\" --type hdd --port "
                   + QString::number(port) + " --device " + QString::number(device) + " --medium \"" + diskPath + "\"").toLatin1());
}
int vbox::MountISO(QString isoPath, QString controlName, int port, int device){
    return system(("\"" + managerPath + "\" storageattach \"" + name + "\" --storagectl \"" +
                   controlName + "\" --type dvddrive --port " + QString::number(port) + " --device " + QString::number(device)
                   + " --medium \"" + isoPath + "\"").toLatin1());
}
int vbox::BootFirst(QString bootDrive){
    return system(("\"" + managerPath + "\" modifyvm \"" + name + "\" --boot1 " + bootDrive).toLatin1());
}
int vbox::SetNetBridge(QString netDriver){
    return system(("\"" + managerPath + "\" modifyvm \"" + name +
                   "\" --nic1 bridged --cableconnected1 on --nictype1 82540EM --bridgeadapter1 \"" + netDriver + "\" --intnet1 brigh1 --macaddress1 auto").toLatin1());
}
int vbox::SetCPU(int number){
    return system(("\"" + managerPath + "\" modifyvm \"" + name + "\" --cpus " + QString::number(number)).toLatin1());
}
int vbox::SetMemory(int memory){
    return system(("\"" + managerPath + "\" modifyvm \"" + name + "\" --memory " + QString::number(memory)).toLatin1());
}
int vbox::SetRemote(bool setting){
    if(setting){
        return system(("\"" + managerPath + "\" modifyvm \"" + name + "\" --vrde on").toLatin1());
    }
    return system(("\"" + managerPath + "\" modifyvm \"" + name + "\" --vrde off").toLatin1());
}
int vbox::SetRemoteConnectSetting(int port){
    return system(("\"" + managerPath + "\" modifyvm \"" + name + "\" --vrdeport " + QString::number(port) + " --vrdeaddress """).toLatin1());
}
int vbox::Start(bool unShown){
    if(unShown){
        return system(("\"" + managerPath + "\"").toLatin1());
    }
    return system(("\"" + managerPath + "\" startvm \"" + name + "\"").toLatin1());
}
int vbox::Stop(){
    return system(("\"" + managerPath + "\" controlvm \"" + name + "\" poweroff").toLatin1());
}
int vbox::Delete(){
    return system(("\"" + managerPath + "\" unregistervm --delete \"" + name + "\"").toLatin1());
}
int vbox::SetDisplayMemory(int memory){
    return system(("\"" + managerPath + "\" modifyvm \"" + name + "\" --vram " + QString::number(memory)).toLatin1());
}
int vbox::InstallGuessAdditions(QString controlName, int port, int device){
    return MountISO("/usr/share/virtualbox/VBoxGuestAdditions.iso", controlName, port, device);
}
int vbox::EnabledAudio(){
    return system(("\"" + managerPath + "\" modifyvm \"" + name + "\" --audio-driver pulse --audiocontroller hda --audioin on --audioout on").toLatin1());
}
int vbox::EnabledClipboardMode(){
    return system(("\"" + managerPath + "\" modifyvm \"" + name + "\" --clipboard-mode bidirectional").toLatin1());
}
int vbox::EnabledDraganddrop(){
    return system(("\"" + managerPath + "\" modifyvm \"" + name + "\" --draganddrop bidirectional").toLatin1());
}
int vbox::ShareFile(QString name, QString path){
    return system(("\"" + managerPath + "\" sharedfolder add \"" + this->name + "\" -name \"" + name + "\" -hostpath \"" + path + "\"").toLatin1());
}
int vbox::SetVBoxSVGA(){
    return system(("\"" + managerPath + "\" modifyvm \"" + name + "\" --graphicscontroller vboxsvga").toLatin1());
}
int vbox::SetMousePS2(){
    return system(("\"" + managerPath + "\" modifyvm \"" + name + "\" --mouse usb").toLatin1());
}
int vbox::SetKeyboardPS2(){
    return system(("\"" + managerPath + "\" modifyvm \"" + name + "\" --keyboard usb").toLatin1());
}
int vbox::OpenUSB(){
    return system(("\"" + managerPath + "\" modifyvm \"" + name + "\" --usbohci on").toLatin1());
}

File: VM/vbox.py
import os
import subprocess

class vbox:
    name = ""
    managerPath = ""
    vboxVersion = ""

    def __init__(self, name: str, managerPath = "VBoxManage") -> None:
        self.name = name
        self.managerPath = managerPath
        self.vboxVersion = subprocess.getoutput("'" + managerPath + "' -v")
    
    def Create(self, type = "Windows7"):
        os.system(("\"" + self.managerPath + "\" createvm --name \""
                   + self.name + "\" --ostype \"" + type +
                   "\" --register"))
        return os.system(("\"" + self.managerPath + "\" modifyvm \""
                   + self.name + "\" --ostype \"" + type +
                   "\" "))
        # vboxmanage modifyvm testvm --ostype

    def CreateDisk(self, path: str, size: int):
        return os.system(("\"" + self.managerPath + "\" createvdi --filename \"" + path + "\" --size \"" + str(size) + "\""))

    def CreateDiskControl(self, controlName = "storage_controller_1"):
        return os.system(("\"" + self.managerPath + "\" storagectl \"" + self.name + "\" --name \"" + controlName + "\" --add ide"))

    def MountDisk(self, diskPath: str, controlName = "storage_controller_1", port = 0, device = 0):
        return os.system(("\"" + self.managerPath + "\" storageattach \"" + self.name +
                   "\" --storagectl \"" + controlName + "\" --type hdd --port "
                   + str(port) + " --device " + str(device) + " --medium \"" + diskPath + "\""))

    def MountISO(self, isoPath: str, controlName = "storage_controller_1", port = 1, device = 0):
        return os.system(("\"" + self.managerPath + "\" storageattach \"" + self.name + "\" --storagectl \"" +
                   controlName + "\" --type dvddrive --port " + str(port) + " --device " + str(device)
                   + " --medium \"" + isoPath + "\""))

    def BootFirst(self, bootDrive: str):
        return os.system(("\"" + self.managerPath + "\" modifyvm \"" + self.name + "\" --boot1 " + bootDrive))

    def SetNetBridge(self, netDriver: str):
        return os.system(("\"" + self.managerPath + "\" modifyvm \"" + self.name +
                   "\" --nic1 bridged --cableconnected1 on --nictype1 82540EM --bridgeadapter1 \"" + netDriver + "\" --intnet1 brigh1 --macaddress1 auto"))

    def SetCPU(self, number: int, cpuNum: int, coreNum: int):
        return os.system(("\"" + self.managerPath + "\" modifyvm \"" + self.name + "\" --cpus " + str(number)))

    def SetMemory(self, memory: int):
        return os.system(("\"" + self.managerPath + "\" modifyvm \"" + self.name + "\" --memory " + str(memory)))

    def SetRemote(self, setting: bool):
        if (setting):
            return os.system(("\"" + self.managerPath + "\" modifyvm \"" + self.name + "\" --vrde on"))
        return os.system(("\"" + self.managerPath + "\" modifyvm \"" + self.name + "\" --vrde off"))

    def SetRemoteConnectSetting(self, port = 5540):
        return os.system(("\"" + self.managerPath + "\" modifyvm \"" + self.name + "\" --vrdeport " + str(port) + " --vrdeaddress """))

    def Start(self, unShown = False):
        if (unShown):
            return os.system(("\"" + self.managerPath + "\" > /tmp/windows-virtual-machine-installer-for-wine-runner-install.log 2>&1"))
        return os.system(("\"" + self.managerPath + "\" startvm \"" + self.name + "\"  > /tmp/windows-virtual-machine-installer-for-wine-runner-install.log 2>&1"))

    def Stop(self):
        return os.system(("\"" + self.managerPath + "\" controlvm \"" + self.name + "\" poweroff"))

    def Delete(self):
        return os.system(("\"" + self.managerPath + "\" unregistervm --delete \"" + self.name + "\""))

    def SetDisplayMemory(self, memory: int):
        return os.system(("\"" + self.managerPath + "\" modifyvm \"" + self.name + "\" --vram " + str(memory)))

    def InstallGuessAdditions(self, controlName = "storage_controller_1", port = 1, device = 0):
        return self.MountISO("/usr/share/virtualbox/VBoxGuestAdditions.iso", controlName, port, device);
    
    def EnabledAudio(self):
        return os.system(("\"" + self.managerPath + "\" modifyvm \"" + self.name + "\" --audio-driver pulse --audiocontroller hda --audioin on --audioout on"))

    def EnabledClipboardMode(self):
        return os.system(("\"" + self.managerPath + "\" modifyvm \"" + self.name + "\" --clipboard-mode bidirectional"))

    def EnabledDraganddrop(self):
        return os.system(("\"" + self.managerPath + "\" modifyvm \"" + self.name + "\" --draganddrop bidirectional"))
    
    def ShareFile(self, name: str, path: str):
        return os.system(("\"" + self.managerPath + "\" sharedfolder add \"" + self.name + "\" -name \"" + self.name + "\" -hostpath \"" + path + "\""))

    def SetVBoxSVGA(self):
        return os.system(("\"" + self.managerPath + "\" modifyvm \"" + self.name + "\" --graphicscontroller vboxsvga"))

    def SetMousePS2(self):
        return os.system(("\"" + self.managerPath + "\" modifyvm \"" + self.name + "\" --mouse usb"))

    def SetKeyboardPS2(self):
        return os.system(("\"" + self.managerPath + "\" modifyvm \"" + self.name + "\" --keyboard usb"))
    
    def OpenUSB(self):
        return os.system(("\"" + self.managerPath + "\" modifyvm \"" + self.name + "\" --usbohci on"))

    def EnabledUEFI(self, status: bool):
        if (status):
            return os.system(("\"" + self.managerPath + "\" modifyvm \"" + self.name + "\" --firmware=efi"))
        return os.system(("\"" + self.managerPath + "\" modifyvm \"" + self.name + "\" --firmware=bios"))

    def AutoInstall(self, iso: str):
        return os.system(("vboxmanage unattended install '" + self.name + "' '--iso=" + iso + "'"))
